server/controllers/donation.js

```diff
@@ -53,7 +48,8 @@ export default {
 
   hasAuthorization(req, res, next) {
     if (req.user.roles.find(r => r === ADMIN_ROLE) ||
-        req.params.donationId === req.user._id) {
+        req.params.donationId === req.user._id ||
+        req.body.donor === req.user._id) {
```

Comment:
This can cause an issue since hasAuthorization is also used in sendEmail. A malicious user can login with any user id. Then he can send put requests to `/donations/:donationId` with body `{ donor: (his userId)}`. This will cause hasAuthorization to pass and will allow him to access this endpoint for any `:donationId`. Is there another way you can do this that doesn't open up this hole?

Reply:
Good point will look into.

Reply:
How do you feel about having a separate authorization function for creating donations? I.e. `canCreateDonation`

Reply:
That sounds good to me.   When you make it I think it will be better to call the function from the `create(req, res)` controller method rather than putting it as a middleware in the routes file.

Reply:
There's already an auth check in the donation controller create method, admin or `donation.donor === user.id`. I think that's ok - admins can create donations for anyone, regular users can only create donations for themselves.

This `hasAuthorization` one looks weird though, `donationId === user.id`!? I think it should look up the donation from the db and check the donor is the same as the user.

Reply:
Made following commit to resolve the above issues:

https://github.com/freeCodeCamp/pantry-for-good/pull/286/commits/ca1ea1665e9e52f62b8fd1f66baeaab37670c42b